### PR TITLE
HAAR-2823 Save audit events to S3 bucket

### PR DESCRIPTION
**Pull request by ma226860**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      .build()
+
+    val jsonBytes = auditEventJsonString.toByteArray(StandardCharsets.UTF_8)
+
+    s3Client.putObject(putObjectRequest, RequestBody.fromBytes(jsonBytes))
+  }
+
+  private fun generateFilename(auditEvent: HMPPSAuditListener.AuditEvent, auditEventJsonString: String): String {
+    val whenDateTime = auditEvent.`when`.atZone(ZoneId.systemDefault()).toLocalDateTime()
+    return "${whenDateTime.year}/${whenDateTime.monthValue}/${whenDateTime.dayOfMonth}/${auditEvent.who}/" +
```
Comment (andrewl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt:37]: This needs to be formatted as "year=${whenDateTime.year}/month=${whenDateTime.monthValue}/day=${whenDateTime.dayOfMonth}/user=${auditEvent.who}/"

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import java.time.ZoneId
+
+@Service
+class AuditS3Client(
+  private val s3Client: S3Client,
+  private val objectMapper: ObjectMapper,
+  @Value("\${aws.s3.auditBucketName}") private val bucketName: String,
+) {
+
+  fun save(auditEvent: HMPPSAuditListener.AuditEvent) {
```
Comment (andrewl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt:21]: What happens in the event of a failure? (temporary, maybe can't connect to the S3 API, in which case we can retry) or permanent (eg there's no when or who fields in the user event?)

Comment (ma226860) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3Client.kt:21]: at the minute the request will just fall over. What's our retry policy? re body validation- that happens well before we get to this point so we don't need to worry about audit events have dirty data

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3ClientTest.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    )
+    val auditEventJsonString = """{"when":"${auditEvent.`when`}","what":"testEvent","who":"testUser","details":{"key":"value"}}"""
+    `when`(objectMapper.writeValueAsString(auditEvent)).thenReturn(auditEventJsonString)
+
+    auditS3Client.save(auditEvent)
+
+    verify(s3Client).putObject(putObjectRequestCaptor.capture(), requestBodyCaptor.capture())
+    val putObjectRequest = putObjectRequestCaptor.value
+    assertThat(putObjectRequest.bucket()).isEqualTo("bucketName")
+    assertThat(putObjectRequest.key()).isEqualTo("2020/12/31/testUser/07c867b1fb88d84fdcb16090619da3da.json")
```
Comment (andrewl) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsauditapi/services/AuditS3ClientTest.kt]: this too should be "year=2020/month=12/day=31/user=testUser/07c867b1fb88d84fdcb16090619da3da.json"

**Review by andrewl — APPROVED** [review]
lgtm

**Pull request merged**